src/go/k8s/controllers/core/redpandacluster_controller.go

```diff
@@ -0,0 +1,423 @@
+package core
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1alpha1 "github.com/vectorizedio/redpanda/src/go/k8s/apis/core/v1alpha1"
+	"github.com/vectorizedio/redpanda/src/go/rpk/pkg/config"
+	"gopkg.in/yaml.v3"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	seedSuffix	= "-seed"
+	dataDirectory	= "/var/lib/redpanda/data"
+	fsGroup		= 101
+
+	debugLevel	= 2
+)
+
+// RedpandaClusterReconciler reconciles a RedpandaCluster object
+type RedpandaClusterReconciler struct {
+	client.Client
+	Log	logr.Logger
+	Scheme	*runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.vectorized.io,resources=redpandaclusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.vectorized.io,resources=redpandaclusters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core.vectorized.io,resources=redpandaclusters/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch;create;update;patch;
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch;create;update;patch;
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// the RedpandaCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.0/pkg/reconcile
+// nolint:funlen // The complexity of Reconcile function will be address in the next version
+func (r *RedpandaClusterReconciler) Reconcile(
+	ctx context.Context, req ctrl.Request,
+) (ctrl.Result, error) {
+	log := r.Log.WithValues("redpandacluster", req.NamespacedName)
+
+	var redpandaCluster corev1alpha1.RedpandaCluster
+	if err := r.Get(ctx, req.NamespacedName, &redpandaCluster); err != nil {
+		log.Error(err, "unable to fetch RedpandaCluster")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var svc corev1.Service
+	if err := r.Get(ctx, types.NamespacedName{Name: redpandaCluster.Name, Namespace: redpandaCluster.Namespace}, &svc); err != nil {
```

Comment:
I think it would be better to maybe link all the resources via labels than force them to have the same name? That e.g. prevents us from adding another one in the future. But I am fine with changing this as a follow up. Just wanted to start the discussion here :)

Reply:
I don't think you can introduce new object e.g. service with the same name, but different labelset. I tried to copy all labels set from the CR to every other object (service, configmap and statefuset).

Reply:
oh no with the same name, the object has to be unique on GVK.